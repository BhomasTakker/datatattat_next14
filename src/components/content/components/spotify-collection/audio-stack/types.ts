import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
