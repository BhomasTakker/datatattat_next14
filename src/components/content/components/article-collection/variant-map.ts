import { gridDisplay5, gridDisplay7 } from "./collections/grid-display/display";
import stackColumns from "./collections/stack-columns/stack-columns";
import stackScroller from "./collections/stack-scroller/stack-scroller";

// we culd do this bette
// type is only different in the renderMethod - JSX.Element & JSX.Element[]
type Variants =
	| typeof stackScroller
	| typeof stackColumns
	| typeof gridDisplay5
	| typeof gridDisplay7;

export enum ArticleCollectionVariants {
	StackScroller = "stack-scroller",
	StackColumns = "stack-columns",
	gridDisplay5 = "grid-display-5",
	gridDisplay7 = "grid-display-7",
}

export const VariantsMap = new Map<ArticleCollectionVariants, Variants>([
	[ArticleCollectionVariants.StackScroller, stackScroller],
	[ArticleCollectionVariants.StackColumns, stackColumns],
	[ArticleCollectionVariants.gridDisplay5, gridDisplay5],
	[ArticleCollectionVariants.gridDisplay7, gridDisplay7],
]);
