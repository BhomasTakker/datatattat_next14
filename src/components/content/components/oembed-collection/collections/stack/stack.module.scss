.root {
	.list {
		display: flex;
		flex-direction: column;
		align-items: center;
		width: 100%;
		li {
			// sm, md, etc
			width: 100%;
		}
	}
}
