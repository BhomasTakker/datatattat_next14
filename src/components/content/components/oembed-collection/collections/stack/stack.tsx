import { OEmbed } from "@/types/data-structures/oembed";
import { ClientOembed } from "../../../content-oembed/client-component";
import styles from "./stack.module.scss";

export type OembedStackProps = {};

const renderOembed = (item: OEmbed) => {
	const { src, html } = item || {};

	return (
		<li key={item.id} data-testid="content-oembed" className={styles.root}>
			{html && <ClientOembed html={html} />}
		</li>
	);
};

const renderMethod = (collection: OEmbed[] = [], props: OembedStackProps) => {
	return (
		<ul className={`${styles.list}`}>
			{collection.map((item) => renderOembed(item))}
		</ul>
	);
};

const oembedStack = {
	styles,
	renderMethod,
};

export default oembedStack;
