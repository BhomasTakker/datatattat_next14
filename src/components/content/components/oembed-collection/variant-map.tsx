// we culd do this bette

import oembedStack from "./collections/stack/stack";

// type is only different in the renderMethod - JSX.Element & JSX.Element[]
type Variants = typeof oembedStack;

export enum OembedCollectionVariants {
	OembedStack = "oembed-stack",
}

export const VariantsMap = new Map<OembedCollectionVariants, Variants>([
	[OembedCollectionVariants.OembedStack, oembedStack],
]);
