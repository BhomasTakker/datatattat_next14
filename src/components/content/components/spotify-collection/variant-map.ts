import spotifyOembed from "./oembed/oembed";

// type is only different in the renderMethod - JSX.Element & JSX.Element[]
type Variants = typeof spotifyOembed;

export enum SpotifyCollectionVariants {
	SpotifyOembed = "spotify-oembed",
}

export const VariantsMap = new Map<SpotifyCollectionVariants, Variants>([
	[SpotifyCollectionVariants.SpotifyOembed, spotifyOembed],
]);
