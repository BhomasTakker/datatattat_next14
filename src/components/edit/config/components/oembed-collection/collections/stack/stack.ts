import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";

export const OEMBED_STACK_CONFIG: InputListProps = {
	id: "stack",
	type: EditInputs.inputList,
	label: "Stack",
	inputs: [
		{
			id: "oembedComponentTitle",
			type: EditInputs.title,
			title: "Oembed Stack Component",
		},
	],
};
