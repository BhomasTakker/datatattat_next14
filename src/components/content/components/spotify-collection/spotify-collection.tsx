import { ComponentProps } from "@/types/component";
import { ClientOembed } from "../oembed-collection/content-oembed/client-component";
import { OEmbed } from "@/types/data-structures/oembed";
import { VariantsMap } from "./variant-map";

type SpotifyCollectionProps = {
	variantType: string;
	items: [];
};

export const SpotifyCollection = ({
	component,
	dataObject,
}: ComponentProps) => {
	const { componentProps } = component;
	const { variantType, ...rest } = componentProps || {};
	const { items = [] } = (dataObject?.data as SpotifyCollectionProps) || {};

	const variantObject = VariantsMap.get(variantType);

	if (!variantObject) {
		return null;
	}

	const { renderMethod, styles } = variantObject;

	return (
		<div className={styles.root} data-testid={variantType}>
			{renderMethod(items, rest)}
		</div>
	);
};
