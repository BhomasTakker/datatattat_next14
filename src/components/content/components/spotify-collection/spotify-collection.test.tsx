import React from "react";
import { render, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import { SpotifyCollection } from "./spotify-collection";
import { SpotifyCollectionVariants } from "./variant-map";
import type { ComponentProps } from "@/types/component";
import type { OEmbed } from "@/types/data-structures/oembed";
import { ContainerHeightOptions } from "../../../page/components/stack/types";

// Mock the variant-map
jest.mock("./variant-map", () => ({
	SpotifyCollectionVariants: {
		SpotifyOembed: "spotify-oembed",
		SpotifyAudioStack: "spotify-audio-stack",
	},
	VariantsMap: new Map([
		[
			"spotify-oembed",
			{
				renderMethod: jest.fn((items: any[], options: any) => (
					<div data-testid="mock-spotify-oembed">
						{items.map((item, index) => (
							<div key={index} data-testid="mock-oembed-item">
								{item.title || item.name}
							</div>
						))}
						<div data-testid="height-option">{options.height}</div>
					</div>
				)),
				styles: {
					root: "mock-oembed-root-styles",
				},
			},
		],
		[
			"spotify-audio-stack",
			{
				renderMethod: jest.fn((items: any[], options: any) => (
					<div data-testid="mock-audio-stack">
						{items.map((item, index) => (
							<div key={index} data-testid="mock-audio-item">
								{item.title || item.name}
							</div>
						))}
						<div data-testid="height-option">{options.height}</div>
					</div>
				)),
				styles: {
					root: "mock-audio-stack-root-styles",
				},
			},
		],
	]),
}));

describe("SpotifyCollection Test Suite", () => {
	const mockSpotifyOembedItems: OEmbed[] = [
		{
			id: "1",
			title: "Test Spotify Album",
			type: "rich",
			version: "1.0",
			provider_name: "Spotify",
			provider_url: "https://spotify.com",
			html: '<iframe src="https://open.spotify.com/embed/album/123"></iframe>',
			width: 300,
			height: 152,
		},
		{
			id: "2",
			title: "Test Spotify Track",
			type: "rich",
			version: "1.0",
			provider_name: "Spotify",
			provider_url: "https://spotify.com",
			html: '<iframe src="https://open.spotify.com/embed/track/456"></iframe>',
			width: 300,
			height: 152,
		},
	];

	const mockSpotifyAudioItems = [
		{
			id: "test-track-1",
			title: "Test Song 1",
			src: "https://spotify.com/track1",
			description: "Test Description 1",
			guid: "guid-track-1",
			variant: "spotify-track",
			media: {
				type: "track",
			},
		},
		{
			id: "test-album-1",
			title: "Test Album 1",
			src: "https://spotify.com/album1",
			description: "Test Album Description",
			guid: "guid-album-1",
			variant: "spotify-album",
			media: {
				type: "album",
			},
		},
	];

	const createMockProps = (
		items: any[],
		variantType: string = SpotifyCollectionVariants.SpotifyOembed,
		additionalProps: any = {}
	): ComponentProps => ({
		component: {
			componentProps: {
				variantType,
				...additionalProps,
			},
		} as any,
		dataObject: {
			data: {
				items,
			},
		},
	});

	beforeEach(() => {
		jest.clearAllMocks();
	});

	afterEach(() => {
		cleanup();
		jest.clearAllMocks();
	});

	describe("Basic Rendering", () => {
		it("should render SpotifyOembed variant correctly", () => {
			const props = createMockProps(
				mockSpotifyOembedItems,
				SpotifyCollectionVariants.SpotifyOembed
			);
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("spotify-oembed")).toBeInTheDocument();
			expect(screen.getByTestId("spotify-oembed")).toHaveClass(
				"mock-oembed-root-styles"
			);
		});

		it("should render SpotifyAudioStack variant correctly", () => {
			const props = createMockProps(
				mockSpotifyAudioItems,
				SpotifyCollectionVariants.SpotifyAudioStack
			);
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("spotify-audio-stack")).toBeInTheDocument();
			expect(screen.getByTestId("spotify-audio-stack")).toHaveClass(
				"mock-audio-stack-root-styles"
			);
		});

		it("should return null when variant is not found", () => {
			const props = createMockProps(
				mockSpotifyOembedItems,
				"invalid-spotify-variant"
			);
			const { container } = render(<SpotifyCollection {...props} />);

			expect(container.firstChild).toBeNull();
		});

		it("should handle undefined variantType", () => {
			const props: ComponentProps = {
				component: {
					componentProps: {},
				} as any,
				dataObject: {
					data: {
						items: mockSpotifyOembedItems,
					},
				},
			};
			const { container } = render(<SpotifyCollection {...props} />);

			expect(container.firstChild).toBeNull();
		});
	});

	describe("Items Rendering", () => {
		it("should render Spotify oembed items using the variant render method", () => {
			const props = createMockProps(
				mockSpotifyOembedItems,
				SpotifyCollectionVariants.SpotifyOembed
			);
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("mock-spotify-oembed")).toBeInTheDocument();
			expect(screen.getAllByTestId("mock-oembed-item")).toHaveLength(2);
			expect(screen.getByText("Test Spotify Album")).toBeInTheDocument();
			expect(screen.getByText("Test Spotify Track")).toBeInTheDocument();
		});

		it("should render Spotify audio items using the variant render method", () => {
			const props = createMockProps(
				mockSpotifyAudioItems,
				SpotifyCollectionVariants.SpotifyAudioStack
			);
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("mock-audio-stack")).toBeInTheDocument();
			expect(screen.getAllByTestId("mock-audio-item")).toHaveLength(2);
			expect(screen.getByText("Test Song 1")).toBeInTheDocument();
			expect(screen.getByText("Test Album 1")).toBeInTheDocument();
		});

		it("should handle empty items array", () => {
			const props = createMockProps(
				[],
				SpotifyCollectionVariants.SpotifyOembed
			);
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("spotify-oembed")).toBeInTheDocument();
			expect(screen.getByTestId("mock-spotify-oembed")).toBeInTheDocument();
			expect(screen.queryAllByTestId("mock-oembed-item")).toHaveLength(0);
		});

		it("should handle missing items property", () => {
			const props: ComponentProps = {
				component: {
					componentProps: {
						variantType: SpotifyCollectionVariants.SpotifyOembed,
					},
				} as any,
				dataObject: {
					data: {},
				},
			};
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("spotify-oembed")).toBeInTheDocument();
			expect(screen.queryAllByTestId("mock-oembed-item")).toHaveLength(0);
		});

		it("should handle undefined dataObject", () => {
			const props: ComponentProps = {
				component: {
					componentProps: {
						variantType: SpotifyCollectionVariants.SpotifyOembed,
					},
				} as any,
				dataObject: undefined as any,
			};
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("spotify-oembed")).toBeInTheDocument();
			expect(screen.queryAllByTestId("mock-oembed-item")).toHaveLength(0);
		});
	});

	describe("Props Passing", () => {
		it("should pass additional props to render method for SpotifyOembed", () => {
			const additionalProps = {
				height: ContainerHeightOptions.LG,
				customProp: "test-value",
			};
			const props = createMockProps(
				mockSpotifyOembedItems,
				SpotifyCollectionVariants.SpotifyOembed,
				additionalProps
			);
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.LG
			);
		});

		it("should pass additional props to render method for SpotifyAudioStack", () => {
			const additionalProps = {
				height: ContainerHeightOptions.SM,
				customProp: "test-value",
			};
			const props = createMockProps(
				mockSpotifyAudioItems,
				SpotifyCollectionVariants.SpotifyAudioStack,
				additionalProps
			);
			render(<SpotifyCollection {...props} />);

			expect(screen.getByTestId("height-option")).toHaveTextContent(
				ContainerHeightOptions.SM
			);
		});

		it("should exclude variantType from props passed to render method", () => {
			const { VariantsMap } = require("./variant-map");
			const mockVariant = VariantsMap.get("spotify-oembed");

			const props = createMockProps(
				mockSpotifyOembedItems,
				SpotifyCollectionVariants.SpotifyOembed,
				{ height: ContainerHeightOptions.MD, customProp: "test" }
			);
			render(<SpotifyCollection {...props} />);

			// Check that renderMethod was called with the correct arguments
			expect(mockVariant.renderMethod).toHaveBeenCalledWith(
				mockSpotifyOembedItems,
				{
					height: ContainerHeightOptions.MD,
					customProp: "test",
				}
			);
		});
	});

	describe("Component Structure", () => {
		it("should render with correct data-testid attribute", () => {
			const props = createMockProps(
				mockSpotifyOembedItems,
				SpotifyCollectionVariants.SpotifyOembed
			);
			render(<SpotifyCollection {...props} />);

			const container = screen.getByTestId("spotify-oembed");
			expect(container).toBeInTheDocument();
		});

		it("should apply correct CSS classes from variant styles", () => {
			const props = createMockProps(
				mockSpotifyAudioItems,
				SpotifyCollectionVariants.SpotifyAudioStack
			);
			render(<SpotifyCollection {...props} />);

			const container = screen.getByTestId("spotify-audio-stack");
			expect(container).toHaveClass("mock-audio-stack-root-styles");
		});
	});

	describe("Error Handling", () => {
		it("should handle null componentProps", () => {
			const props: ComponentProps = {
				component: {
					componentProps: null,
				} as any,
				dataObject: {
					data: {
						items: mockSpotifyOembedItems,
					},
				},
			};
			const { container } = render(<SpotifyCollection {...props} />);

			expect(container.firstChild).toBeNull();
		});

		it("should handle missing component property", () => {
			const props: ComponentProps = {
				component: {} as any,
				dataObject: {
					data: {
						items: mockSpotifyOembedItems,
					},
				},
			};
			const { container } = render(<SpotifyCollection {...props} />);

			expect(container.firstChild).toBeNull();
		});

		it("should handle variant with missing styles", () => {
			const { VariantsMap } = require("./variant-map");
			const originalVariant = VariantsMap.get("spotify-oembed");

			// Mock variant with empty styles object
			VariantsMap.set("spotify-oembed", {
				renderMethod: originalVariant.renderMethod,
				styles: { root: "" },
			});

			const props = createMockProps(
				mockSpotifyOembedItems,
				SpotifyCollectionVariants.SpotifyOembed
			);

			// Should render without error
			const { container } = render(<SpotifyCollection {...props} />);
			expect(container.firstChild).toBeInTheDocument();

			// Restore original variant
			VariantsMap.set("spotify-oembed", originalVariant);
		});
	});

	describe("Snapshot Testing", () => {
		it("should match snapshot for SpotifyOembed variant", () => {
			const props = createMockProps(
				mockSpotifyOembedItems,
				SpotifyCollectionVariants.SpotifyOembed
			);
			const { container } = render(<SpotifyCollection {...props} />);

			expect(container.firstChild).toMatchSnapshot();
		});

		it("should match snapshot for SpotifyAudioStack variant", () => {
			const props = createMockProps(
				mockSpotifyAudioItems,
				SpotifyCollectionVariants.SpotifyAudioStack
			);
			const { container } = render(<SpotifyCollection {...props} />);

			expect(container.firstChild).toMatchSnapshot();
		});

		it("should match snapshot for null case", () => {
			const props = createMockProps(mockSpotifyOembedItems, "invalid-variant");
			const { container } = render(<SpotifyCollection {...props} />);

			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
