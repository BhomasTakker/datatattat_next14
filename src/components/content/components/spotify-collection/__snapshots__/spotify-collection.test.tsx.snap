// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SpotifyCollection Test Suite Snapshot Testing should match snapshot for SpotifyAudioStack variant 1`] = `
<div
  class="mock-audio-stack-root-styles"
  data-testid="spotify-audio-stack"
>
  <div
    data-testid="mock-audio-stack"
  >
    <div
      data-testid="mock-audio-item"
    >
      Test Song 1
    </div>
    <div
      data-testid="mock-audio-item"
    >
      Test Album 1
    </div>
    <div
      data-testid="height-option"
    />
  </div>
</div>
`;

exports[`SpotifyCollection Test Suite Snapshot Testing should match snapshot for SpotifyOembed variant 1`] = `
<div
  class="mock-oembed-root-styles"
  data-testid="spotify-oembed"
>
  <div
    data-testid="mock-spotify-oembed"
  >
    <div
      data-testid="mock-oembed-item"
    >
      Test Spotify Album
    </div>
    <div
      data-testid="mock-oembed-item"
    >
      Test Spotify Track
    </div>
    <div
      data-testid="height-option"
    />
  </div>
</div>
`;

exports[`SpotifyCollection Test Suite Snapshot Testing should match snapshot for null case 1`] = `null`;
